samples/nrf9160/modem_shell/src/gnss/gnss_shell.c

```diff
@@ -0,0 +1,666 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <stdlib.h>
+
+#include <zephyr.h>
+#include <shell/shell.h>
+#include <getopt.h>
+
+#include "gnss.h"
+
+/* Sets the global shell variable, must be called before calling any GNSS
+ * module functions.
+ */
+#define GNSS_SET_GLOBAL_SHELL()	(shell_global = shell)
+
+extern const struct shell *shell_global;
+
+static bool gnss_running;
+
+static int print_help(const struct shell *shell, size_t argc, char **argv)
+{
+	int ret = 1;
+
+	if (argc > 1) {
+		shell_error(shell, "%s: subcommand not found", argv[1]);
+		ret = -EINVAL;
+	}
+
+	shell_help(shell);
+
+	return ret;
+}
+
+static int cmd_gnss(const struct shell *shell, size_t argc, char **argv)
+{
+	GNSS_SET_GLOBAL_SHELL();
```

Comment:
It seems that the sample is limited to UART shell anyway (looking at h the rest of the code). Perhaps it'd be better to set this global shell pointer once during initialization instead of calling this macro in every single command handler?

Reply:
That would be great, but I don't know where we could do that. That global shell pointer would need to be set when the shell is initialized, but I haven't figured out a way to do that.

Reply:
We have this in the code somewhere:
`shell_global = shell_backend_uart_get_ptr();`
But all shell command functions are setting shell_global in the beginning to make sure the pointer is available because we haven't found a proper place to initialize it just once. In main() shell doesn't seem to be available.
@rlubos So any ideas are more than welcome.

Reply:
> In main() shell doesn't seem to be available

I was just about to suggest to do this at the beginning of `main()`, why wouldn't shell be avaialable there? It should be initialized before the application enters `main()`, isn't it enough to simply `#include <shell/shell_uart.h>` to use the appropriate function?

Reply:
Earlier we had problems with main() not being called before the shell became usable. I think this happened because LTE LC blocked until LTE was connected (LTE LC auto connect was used) and this prevented main() from being called, however the shell was initialized already before LTE LC. We don't use CONFIG_LTE_AUTO_INIT_AND_CONNECT anymore, so this is probably not an issue anymore.

Reply:
Another solution could be using `SYS_INIT` macro with some initializing function - the init priority could be finetuned so that it's called just after SHELL gets started.

Reply:
Using` shell_backend_uart_get_ptr()` in the beginning of `main()` seems to work well enough.
If I use the pointer for `shell_print()` immediately after getting the pointer, it prints:
`WARNING: A print request was detected on not active shell backend.`

Using `shell_print()` with given pointer works fine few lines below after:
`err = nrf_modem_lib_init(NORMAL_MODE);`
So, looks like shell backend has some initializations ongoing early on in `main()`.

Reply:
Another reason this might have worked in a worse manner earlier is that we used to use `CONFIG_LTE_AUTO_INIT_AND_CONNECT=y` but not anymore.